### PR TITLE
ENH: Speed up reading of small buffers

### LINKED ISSUE #12234: ENH: Speed up FIF IO
From @britta-wstnr use case if a FIF file is written one sample per data chunk:
```
import mne
import numpy as np
data = np.random.RandomState(0).randn(32, 5000 * 8 * 60)
info = mne.create_info(32, 5000., "mag")
raw = mne.io.RawArray(data, info)
raw.save("test_default_raw.fif", overwrite=True)
raw.save("test_1samp_raw.fif", buffer_size_sec=1. / raw.info["sfreq"], overwrite=True)  # takes ages
```
you get a much slower read:
```
$ time python -c 'import mne; mne.io.read_raw_fif("test_default_raw.fif").load_data()'
real	0m1.496s
$ time python -c 'import mne; mne.io.read_raw_fif("test_1samp_raw.fif").load_data()'
real	0m27.146s
```
in principle if we know all tags are contiguous on disk, we can use a structured `np.dtype` consisting of tag header plus tag data to read in all necessary tags at once then populate our output data from the tag data (discarding the tag header), avoiding the costly loop.

I'll try to take a look at some point!

### PR DESCRIPTION
**Pull request by larsoner**
Closes #12234

@britta-wstnr at the end of the day I don't think the `numpy.dtype` I considered in #12234 will be tractable/worth it as:

1. there can be corrupt files (uncommon but do exist) and we can only find them by actually trying to read each buffer
2. there can be acquisition skips and these need to be handled properly (currentlny easily handled in a loop... not easily if we try to vectorize)
3. the single-sample case is pretty rare and sounds like there should be a fix at the acq end soon (?)
4. there is a nice workaround of reading the files *once* with MNE-Python then writing them out with `buffer_size_sec=1.`, then the I/O is no longer slow

Okay with you to close the associated issue when this is merged?

In the meantime this PR cleans up and speeds up our tag code, on my system it:

1. Speeds up the single-sample-buffer snippet from #12234 by a factor of about 2
2. Speeds up `pytest mne/io/fiff` from `23.79s` to `20.46s`
3. Simplifies the `Tag` logic and functions, especially involving *when* to `fid.seek` to a new position in a file (it's explicit now rather than implicit)
4. Removes a completely unused `copy_tree` which is already in a private namespace (`mne._fiff`) so no need to deprecate, can just remove
5. Prefers using explicit file positions to using `fid.tell` because this function was surprisingly a source of slowdowns in the single-sample buffer case

### CODE REVIEW

File: mne/_fiff/open.py [hunk 1]
```diff
@@ -131,6 +137,7 @@ def fiff_open(fname, preload=False, verbose=None):
         raise
 
 
+# @profile
```
Comment (drammock) [inline mne/_fiff/open.py:140]: cruft

```suggestion
```

File: mne/_fiff/open.py [hunk 2]
```diff
@@ -309,7 +315,7 @@ def _show_tree(
         for k, kn, size, pos, type_ in zip(kinds[:-1], kinds[1:], sizes, poss, types):
             if not tag_found and k != tag_id:
                 continue
-            tag = Tag(k, size, 0, pos)
+            tag = Tag(k, size, 0, -1, pos)
```
Comment (drammock) [inline mne/_fiff/open.py:318]: is this order correct?  `Tag` params order is `kind`, `type`, `size`, `next`, `pos`, looks like `type` and `size` are reversed here

File: mne/_fiff/tag.py [hunk 3]
```diff
@@ -456,15 +441,17 @@ def read_tag(fid, pos=None, shape=None, rlims=None):
         Note that data are assumed to be stored row-major in the file. Only to
         be used with data stored as a vector (not implemented for matrices
         yet).
+    advance : bool
+        If True (default), advance to next tag position after reading.
+
+        .. versionadded:: 1.7
```
Comment (drammock) [inline mne/_fiff/tag.py:447]: according to the diff, this new param is not actually used?

File: mne/commands/tests/test_commands.py [hunk 4]
```diff
@@ -100,13 +100,22 @@ def test_compare_fiff():
...
         mne_show_fiff.run()
+    bad_fname = tmp_path / "test_bad_raw.fif"
+    with open(bad_fname, "wb") as fout:
+        with open(raw_fname, "rb") as fin:
+            fout.write(fin.read(100000))
+    with pytest.warns(RuntimeWarning, match="Invalid tag"):
+        lines = show_fiff(str(bad_fname), output=list)
+    last_line = lines[-1]
+    assert ">>>>BAD" in last_line
+    assert "302  = FIFF_EPOCH (734412b >f4)" in last_line
```
Comment (larsoner) [inline mne/commands/tests/test_commands.py:118]: Nice catch on the order above @drammock, this line used to be:
```
302  = FIFF_EPOCH (734412b ?1073741828?) >>>>>>>>>>>>>>>>>>>>BAD'
```
which is wrong, now it's correct:
```
302  = FIFF_EPOCH (734412b >f4) >>>>>>>>>>>>>>>>>>>>BAD
```

**Review by drammock — APPROVED** [review]

**Pull request merged**